src/main/java/com/icthh/xm/ms/entity/service/ContentService.java

```diff
@@ -1,29 +1,128 @@
 package com.icthh.xm.ms.entity.service;
 
+import com.amazonaws.util.IOUtils;
 import com.icthh.xm.commons.permission.annotation.FindWithPermission;
 import com.icthh.xm.commons.permission.annotation.PrivilegeDescription;
 import com.icthh.xm.commons.permission.repository.PermittedRepository;
+import com.icthh.xm.ms.entity.domain.Attachment;
+import com.icthh.xm.ms.entity.domain.AttachmentStoreType;
 import com.icthh.xm.ms.entity.domain.Content;
-import com.icthh.xm.ms.entity.repository.search.PermittedSearchRepository;
+import com.icthh.xm.ms.entity.domain.spec.AttachmentSpec;
+import com.icthh.xm.ms.entity.repository.AttachmentRepository;
+import com.icthh.xm.ms.entity.repository.ContentRepository;
+import com.icthh.xm.ms.entity.repository.backend.S3StorageRepository;
+import com.icthh.xm.ms.entity.service.dto.S3ObjectDto;
+import com.icthh.xm.ms.entity.service.dto.UploadResultDto;
 import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
+import java.io.IOException;
 import java.util.List;
 
+@Slf4j
 @Service
 @Transactional
 @RequiredArgsConstructor
 public class ContentService {
 
-    private final PermittedRepository permittedRepository;
+    private static final String FILE_NAME_SEPARATOR = "::";
 
-    private final PermittedSearchRepository permittedSearchRepository;
+    private final PermittedRepository permittedRepository;
+    private final ContentRepository contentRepository;
+    private final S3StorageRepository s3StorageRepository;
 
+    @Transactional(readOnly = true)
     @FindWithPermission("CONTENT.GET_LIST")
     @PrivilegeDescription("Privilege to get all the contents")
     public List<Content> findAll(String privilegeKey) {
         return permittedRepository.findAll(Content.class, privilegeKey);
     }
 
+    public Attachment save(AttachmentSpec spec, Attachment attachment, Content content) {
```

Comment:
attachment has typeKey, and spec should be fetched, not passed as parameter to this level